Geolocation foreground listener

vividos opened the pull request:
### Description of Change ###

This is an initial implementation of a foreground-only listener for Location changes. It consists of initial work from #1043 (done by Alex Reich), the API proposed in #1447 (Spec by me), and code from James Montemagno's Geolocator-Plugin.

### Bugs Fixed ###

- Fixes #290
- Closes #1043
- Closes #1447

### API Changes ###

See Spec in #1447

### Behavioral Changes ###

None.

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of `main` at time of PR
- [x] Changes adhere to coding standard
- [ ] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc))

Tests not added yet, as well as mdoc docs.

vividos commented:
@mattleibow  @jamesmontemagno @Redth Any comments about the current state of the PR are very welcome! Will do the UWP implementation next. Help for WPF and Tizen platforms are welcome!

mattleibow commented:
This is looking really nice. I'll have a closer look, but so far it all makes sense.

Thanks for this PR!

vividos commented:
Thanks for the feedback! One thing that occurred to me was that the code is using `GeolocationRequest.Timeout` as the minimum timeout, which is a bit misleading. Also if `StopListeningForegroundAsync` needs a return type at all.

vividos commented:
I rebased my PR on top of current main, in order to keep the PR current and to check if it still builds (which I guess).

vividos commented:
@jamesmontemagno I saw you're added the needs-more-discussion tag. Let's discuss! I can't wait to get this PR into Xamarin.Essentials! :-D

vividos commented:
One thing that I think doesn't match exactly is the re-use of the `GeolocationRequest` class to pass options to `StartListeningForegroundAsync()`; here, I'm reusing `request.Timeout` for the minimum time. I could add a new property `ListeningMinimumTime` to the request object to clarify that it's for the listening part of `Geolocation`. What do you think?

vividos commented:
I also just rebased the PR on top of current main, just to be current.

janusw commented:
> I can't wait to get this PR into Xamarin.Essentials! :-D

Me too :D

+1 for merging this! 👍

vividos commented:
I rebased the commits onto the latest main branch (that's also tagged with 1.7.0) squashed together some commits that fixed bug in the previous commits and added the new `ListeningRequest` class.

vividos commented:
It seems there won't be any development in Essentials anymore - closing this PR.
